keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,737 @@
+     configuration; if the user wants IPv6 addresses to be removed,
+     they must configure kubelet to request an obligatorily
+     single-stack IPv4 cluster.
+
+     (Of course, they only actually need to do that if they're running
+     their obligatorily single-stack IPv4 cluster on
+     actually-dual-stack hosts, which they presumably aren't.)
+
+The rest of this assumes the last approach.
+
+#### Proposal: `--node-ips`, Primary Node IP, Secondary Node IP
+
+To allow more consistent behavior without breaking existing users, we
+will deprecate `--node-ip` and replace it with a new `--node-ips`
+argument with more powerful and more-consistent semantics, and change
+the way that we generate, filter, and sort `Node.Status.Addresses` to
+more reliably choose and indicate the Primary Node IP, and (when
+available) the preferred Secondary Node IP of the opposite IP family.
+
+`--node-ips` can be either a single element describing the node IP
+configuration for a single family (indicating an
+obligatorily-single-stack node), or else a pair of elements describing
+both the IPv4 and IPv6 configurations. Each element can be:
+
+  - a specific IP address, eg, `10.0.0.1` or `fd01::1`, indicating
+    that the node _must_ have that IP address for that IP family
+
+  - the string "`ipv4`" or "`ipv6`", indicating that the node can have
+    _any_ address of that family (but doesn't necessarily have to).
+
+The default value is `ipv4,ipv6`, which allows for any of single-stack
+IPv4, single-stack IPv6, or dual stack (with IPv4 primary). By
+contrast, `--node-ips ipv4` would mean single-stack IPv4 only (and
+would cause IPv6 addresses to be removed from `Node.Status.Addresses`)
+(Note that there is no way to say "dual-stack only".)
+
+```
+<<[UNRESOLVED optional-vs-required ]>>
+
+If we wanted to be able to have both "optionally dual-stack" and
+"mandatory dual-stack" we could have separate "optional ipv4/6" and
+"required ipv4/6" flags, but that seems to result in a lot of
+complication for not much benefit.
+
+<<[/UNRESOLVED]>>
+```
+
+If `--node-ips` is not specified but `--node-ip` is, then `--node-ips`
+is defaulted to an equivalent value as follows:
+
+  - If `--node-ip` is `0.0.0.0` then `--node-ips` is set to `ipv4`.
+
+  - If `--node-ip` is `::` then `--node-ips` is set to `ipv6`.
+
+  - Otherwise (when `--node-ip` is a specific IP address) `--node-ips`
+    is set to the same value as `--node-ip`.
+
+It is an error to pass both `--node-ip` and `--node-ips`.
+
+As discussed under [Implementation](#implementation),
+`Node.Status.Addresses` will then be set correctly to reflect this.
+Once `Node.Status.Addresses` has been set, the rule for actually
+finding the Primary Node IP is still the same as it used to be; we
+will just adjust `Node.Status.Addresses` to ensure that someone
+following the traditional rule will find the IP we want them to.
+
+There is now also a "Secondary Node IP", which is the first
```

Comment:
Hm? It's just using the same rule for the secondary IP as for the primary IP. Historically, the primary node IP is either the first Internal IP, or else the first External IP if there are no Internal IPs. The secondary node IP uses the same rule, except only looking at the IPs of the opposite family from the primary IP.

We _could_ use a different rule for the secondary IP since there's no backward compat to worry about, but there didn't seem to be any reason to. (And I didn't want to change the rule for the primary IP because some external code may get confused if the rule changes.)